onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -329,6 +331,20 @@ Status ModelBuilder::RegisterModelOutputs() {
                              "The output of graph is not registered [", output_name, "]");
     }
 
+    // Since for now all the shapes are deterministic for NNAPI, it's impossible we can have unknown output shape
+    const auto* shape_proto = node_arg->Shape();
+    ORT_RETURN_IF_NOT(shape_proto != nullptr, "shape_proto cannot be null for output: ", output_name);
+    if (shape_proto->dim_size() == 0) {
+      // In NNAPI scalar output must have {1} shape
+      const auto& output_shape = shaper_[output_name];
+      ORT_RETURN_IF_NOT(output_shape.size() == 1 && output_shape[0] == 1,
+                        "scalar output [", output_name, "] must have {1} shape, ",
+                        " actual shape, ", Shape2String(output_shape));
+
+      // Record the scalar output
+      nnapi_model_->AddScalarOutput(output_name);
```

Comment:
nit: maybe say that we record it so that we can undo the shape change when returning values to ORT